src/Dependencies/PooledObjects/PooledDisposer.cs

```diff
@@ -3,12 +3,10 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.PooledObjects
 {
-    [NonCopyable]
```

Comment:
not available in shared layer.  so pulled in through a partial part in a higher layer.

Reply:
This was previously rejected from the compiler layer in https://github.com/dotnet/roslyn/pull/37543.

Reply:
hrm... the reasoning doesn't seem to hold up afaict:

> Yeah, let's not do this on the compiler side. I think it's a bad pattern for a couple reasons:

> It trades risk of leaks, which don't affect program semantics, for potential double frees, which do affect program semantics.

Experience on IDE side hasn't shown this to be an issue.  Whereas leaks are *exceedingly* simple to introduce (lol, i just did that here :D).

> It introduces a try-finally, which is very expensive in the CLR.

Looking at teh compiler, it already extensively uses try/finally for exactly this purpose.  Indeed, we use try/finally all over the place (or `using`) for these same patterns.  It seems the simplicy and elegance of the cleanup is definitely something desired all over the place.  Unclear why this helper would be different.

> It introduces a virtual call instead of a direct call for freeing.

Hrmm... that seems untrue.  THis is a struct disposer.  I do see a constrained call.  But a constrained call on a struct is not going to be a vcall at runtime right?

--

Finally, while i'd say there was potential reason to be concerned initially (new pattern and all that), it's proven itself out *Extensively* on the IDE side, with thousands of usages in our code without issue.  It was also the pattern we took with things like DisposableResetPoint.  So i'd say we now have a ton of experience under our belt to feel that this is ok.

The benefit is def clearly worth it as well.  I apparently can't avoid leaking memory :D

Reply:
I'd rather get a read on what the entire team thinks about this before accepting the pooling change, given that it was previously rejected.

Reply:
Sure.  I have no issue with that.  I'm just giving my reasoning why for why that initial rejection seems suspect now.

Reply:
I'm completely fine if the compiler team as a whole decides to continue to disallow this. And would rather bank the perf win here than have to spend a lot of extra time on this specific PR, if this ends up being a protracted conversation.

My personal view mostly aligns with Cyrus's though. I always feel suspicious when we have to avoid coding patterns which are easier to use correctly because of theoretical perf concerns. (where theoretical means we didn't measure it and observe that it results in an unacceptable regression in perf.)

I would agree that double-free is not a risk here. The places where double-free is a risk are where we have to acquire the thing in one method and plunk it through a bunch of helpers before Free'ing it separately on a bunch of different code paths. We are not at a risk of passing this to a component which sees this as an abstract IDisposable and double-frees it because the IDisposable contract says it's allowed to.

I also wouldn't mind seeing if we could establish a checked acquire-release pattern, which doesn't necessitate the IDisposable interface or a try/finally, in the language through some means. A pattern where the expectation is that we don't care if we fail to release the thing when an exception occurs. For example, by adding an analyzer which performs a flow analysis, and a pattern for type declarations to indicate that the method which acquires the thing must also release it.

Reply:
i'm also pretty skeptical that try/finally is a problem outside of maybe the most intense hotspots.  The parser (which we need to be *staggeringly* fast) is chock full of this try/finally/dispose pattern, and yet never ever itself shows up as any sort of perf concern in any of the many core perf scenarios we care deeply about.  I would expect that if this was a problem, we'd def notice, as we happily/freely add try/finally/dispose there and haven't seen even a blip on any radar anywhere.

In my opinion, we should code with clear patterns, and *when* we have data indicating there is a perf issue, we judiciously fix up those precise areas, with the right documentation explaining the issue.

So, in a case like this, we win an absolute ton just by the red->green transition.  And any sort of worry about a loss from try/finally is completely negligible in comparison.